src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkActionFactory.java

```diff
@@ -220,29 +221,100 @@ public void doNothing(String mnemonic, Object inputs) throws EvalException {
   static final GeneratedMessage.GeneratedExtension<ExtraActionInfo, SpawnInfo> SPAWN_INFO =
       SpawnInfo.spawnInfo;
 
+  private String progressMessageForSymlink(
+      Object /* String or None */ progressMessage, Artifact output) {
+    if (progressMessage != Starlark.NONE) {
+      // Should have been verified by Starlark before this function is called.
+      return (String)progressMessage;
+    }
+
+    return "Creating symlink " + output.getRootRelativePathString();
+  }
+
+  /** Return whether exactly one of the passed object is not {@link Starlark.NONE}. */
+  private boolean exactlyOneNotNone(Object first, Object... others) {
```

Comment:
This is a bit of an overgeneralization, but ok. I would have just used a simple if statement, i.e:

```
if ((targetFile == Starlark.NONE) == (targetPath == Starlark.NONE)) {
}
```